packages/dependencies-hierarchy/src/index.ts

```diff
@@ -223,81 +223,112 @@ function getAllDirectDependencies (projectSnapshot: ProjectSnapshot) {
 
-  if (!deps) return []
+    const deps = opts.includeOptionalDependencies === false
+      ? opts.currentPackages[parentId].dependencies
+      : {
+        ...opts.currentPackages[parentId].dependencies,
+        ...opts.currentPackages[parentId].optionalDependencies,
+      }
 
-  const getChildrenTree = getTree.bind(null, {
-    ...opts,
-    currentDepth: opts.currentDepth + 1,
-  })
+    if (!deps) return result
 
-  const peers = new Set(Object.keys(opts.currentPackages[parentId].peerDependencies || {}))
-  const result: PackageNode[] = []
-  Object.keys(deps).forEach((alias) => {
-    const { packageInfo, packageAbsolutePath } = getPkgInfo({
-      alias,
-      currentPackages: opts.currentPackages,
-      modulesDir: opts.modulesDir,
-      peers,
-      ref: deps[alias],
-      registries: opts.registries,
-      skipped: opts.skipped,
-      wantedPackages: opts.wantedPackages,
+    const getChildrenTree = getTreeInner.bind(null, {
+      ...opts,
+      currentDepth: opts.currentDepth + 1,
     })
-    let circular: boolean
-    const matchedSearched = opts.search && opts.search(packageInfo)
-    let newEntry: PackageNode | null = null
-    if (packageAbsolutePath === null) {
-      circular = false
-      newEntry = packageInfo
-    } else {
-      const relativeId = refToRelative(deps[alias], alias) as string // we know for sure that relative is not null if pkgPath is not null
-      circular = keypath.includes(relativeId)
-      const dependencies = circular ? [] : getChildrenTree(keypath.concat([relativeId]), relativeId)
 
-      if (dependencies.length) {
-        newEntry = {
-          ...packageInfo,
-          dependencies,
-        }
-      } else if (!opts.search || matchedSearched) {
+    const peers = new Set(Object.keys(opts.currentPackages[parentId].peerDependencies || {}))
+
+    Object.keys(deps).forEach((alias) => {
+      const { packageInfo, packageAbsolutePath } = getPkgInfo({
+        alias,
+        currentPackages: opts.currentPackages,
+        modulesDir: opts.modulesDir,
+        peers,
+        ref: deps[alias],
+        registries: opts.registries,
+        skipped: opts.skipped,
+        wantedPackages: opts.wantedPackages,
+      })
+      let circular: boolean
+      const matchedSearched = opts.search && opts.search(packageInfo)
+      let newEntry: PackageNode | null = null
+      if (packageAbsolutePath === null) {
+        circular = false
         newEntry = packageInfo
+      } else {
+        const relativeId = refToRelative(deps[alias], alias) as string // we know for sure that relative is not null if pkgPath is not null
+        circular = keypath.includes(relativeId)
+
+        let dependencies: PackageNode[]
+        if (circular) {
+          dependencies = []
+        } else {
+          const cached = dependenciesCache.get(packageAbsolutePath)
+
+          if (cached && !cached.circular) {
```

Comment:
as far as I understand, if circular is true, the cache is never used. So we just need to cache dependencies if circular is false.
```suggestion
          if (cached) {
```